### ISSUE TITLE
Thoughts on a DbCommandInterceptor to skip empty queries

### ISSUE THREAD

**Issue by rctay** [dotnet/efcore#29768]
## Ask a question

Hi, I wrote a DbCommandInterceptor to skip queries that would return an empty result set, and would like some opinions/feedback about it.

### Background
Let's say we have this EFcore query:

```csharp
public IEnumerable<Stuff> GetStuffBySourceId(IEnumerable<string> sourceIds)
{
    sourceIds = sourceIds.ToImmutableHashSet();

    var query = _dbContext.Stuff
        .Include(stuff => stuff.Parents)
        .Include("Parents.ParentStuff")
        .Where(stuff => !stuff.IsDeleted &&
                           // lots more other conjunctions... &&
                           sourceIds.Contains(category.SourceId));
    return query.AsEnumerable();
}
```

When sourceIds is empty, then the SQL (when hooked up with Microsoft SQL Server) generated looks something like

```sql
      SELECT [m].[Column1], ...
      FROM [acme].[Stuff] AS [m]
      LEFT JOIN (
          ...
      ) AS [t] ON [m].[ParentId] = [t].[ParentId]
      WHERE 0 = 1  --this bit here
      ORDER BY [m].[Id], [t].[Id], [t].[ParentStuffId]
```

By hooking on to EF, this means we don't have to add an empty check whenever we do a .Where(). Were the empty check added, then subsequent authors making code changes to the `.Where()` clause (eg. using `||` instead of the strict `&&`) would also have to remember to remove/change the empty check. With the lifecycle hook approach, this risk is reduced.

### Interceptor

Here's what the interceptor looks like:

```C#
using System.Data.Common;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore.Diagnostics;
using AcmeApp.DbInterceptors.DbDataReaders;

namespace AcmeApp.DbInterceptors
{
    /// <summary>
    ///     Skips a DB round-trip for queries that are always empty (eg. WHERE 0 = 1) by getting
    ///     EFCore to immediately return an empty result.
    /// </summary>
    public class SkipEmptyQueriesInterceptor : DbCommandInterceptor
    {
        public override InterceptionResult<DbDataReader> ReaderExecuting(DbCommand command, CommandEventData eventData, InterceptionResult<DbDataReader> result)
        {
            if (!command.CommandText.Contains("WHERE 0 = 1\r\n"))
            {
                    return base.ReaderExecuting(command, eventData, result);
            }

            command.CommandText = "-- skipping DB call as always empty\r\n" + command.CommandText;
            return InterceptionResult<DbDataReader>.SuppressWithResult(new EmptyDbDataReader());
        }

        public override ValueTask<InterceptionResult<DbDataReader>> ReaderExecutingAsync(DbCommand command, CommandEventData eventData, InterceptionResult<DbDataReader> result,
            CancellationToken cancellationToken = new CancellationToken())
        {
            // omitted WLOG
        }
    }
}
```

To avoid making this issue too long, the source code for [EmptyDbDataReader is here](https://gist.github.com/rctay/f8fc4d85b25e40a958c9bd48e03d7427).

### Questions

1. Which part of EFCore does the string SQL generation? I'm interested in looking at the implementation, eg. which line separators are used (`\r\n`), is the line sep used dependent on platform, etc.
2. A lifecycle hook somewhere earlier in the EFCore lifecycle would seem ideal—for example, after the expressions has been evaluated, and still in the "AST"/tree phase before converting to a string/SQL.  Is there such a hook?

   If this form/data structure were available, then I could have better guarantees on the WHERE expression without worrying too much in string-land, like "are there further expressions after the 0=1?". The check on the AST might look like `Where.length == 1 && Where[0].Operator == `=` && Where[0].Operand1 = '0' && Where[0].Operand2 == '1'` (being imaginative about the exact fields here.)

   a. I saw that there was a IDbCommandTreeInterceptor, but that seems to be part of EF and not EFCore, so it can't be used, is that right?
   b. Even if I DbCommandTreeInterceptor were available in EFCore, it only allows command tree transformation, and doesn't have something similar to SuppressWithResult and couldn't be used here—or is there an equivalent for IDbCommandTreeInterceptor?

### Provider and version information

EF Core version: 7.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6.0
Operating system: Windows
IDE: Rider

**Comment by roji** [comment]
As always with SQL rewriting techniques, be very careful... In the general case, just checking for further expressions after `WHERE 0 = 1` isn't enough; the WHERE clausecould be present in some subquery, without necessarily implying that the *entire* query returns no results just because that subquery does. If that happens one day and your interceptor incorrectly skips the query, you'll get wrong results.

Reliably figuring out that a query returns no results is a non-trivial tasks, with nothing-returning subqueries propagating up the query tree in various ways. This is something we could optimize in EF's query pipeline, but I'm not sure it's worth the effort given how much it actually occurs and how much execution of the query at the database actually costs (processing at the database is very likely very cheap, as the database recognizes no rows can come out). It's also pretty trivial to work around in user code by doing an up-front check (though I get your point about this shifting responsibility to the dev).

> Which part of EFCore does the string SQL generation? I'm interested in looking at the implementation, eg. which line separators are used (\r\n), is the line sep used dependent on platform, etc.

That would be [QuerySqlGenerator](https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Query/QuerySqlGenerator.cs), with SQL Server-specific customizations in [SqlServerQuerySqlGenerator](https://github.com/dotnet/efcore/blob/main/src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs) - though I'm not sure that's relevant here. Are you seeing any particular issues with newlines?

> A lifecycle hook somewhere earlier in the EFCore lifecycle would seem ideal—for example, after the expressions has been evaluated, and still in the "AST"/tree phase before converting to a string/SQL. Is there such a hook?

EF's query pipeline is open to extension - you can insert visitors inside it. Since your visitor would need to access parameter values (i.e. to check whether the array is empty), you'd extend [SqlServerParameterBasedSqlProcessor](https://github.com/dotnet/efcore/blob/main/src/EFCore.SqlServer/Query/Internal/SqlServerParameterBasedSqlProcessor.cs) and add your visitor. You would effectively need to implement what I wrote above, assuming you want to handle the general case (recursively propagating the fact that the query returns nothing up the tree).

However, as you noted above this only allows transforming the tree and not suppressing the result. However, your visitor could add a special tag to the query (just like EF's TagWith) which your command interceptor could recognize, causing it to suppress.

To summarize, I'd really recommend making sure the effort here is worth it - there's a pretty strong possibility of it being a premature optimization, with the empty-returning queries costing very little and the (general) implementation complexity being very high (so potential for tricky bugs).
